File: Xamarin/Xamarin.Native/Xamarin.Native.Droid/Activities/FileAccessTestActivity.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Java.Lang;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xamarin.Services;

namespace Xamarin.Native.Droid.Activities
{
    [Activity(Label = "FileAccessTestActivity")]
    public class FileAccessTestActivity : Activity
    {
        string fileName = "testFile.txt";
        int digits = 10000;
        static string contentToWrite;
        Stopwatch stopwatch;

        Button writeButton;
        Button readButton;
        TextView timeLabel;
        TextView resultView;
        ProgressDialog progressDialog;
        FileAccessTestService fileAccessService;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.FileAccessTest);
            Title = GetString(Resource.String.menuFileAccess);

            timeLabel = FindViewById<TextView>(Resource.Id.timeLabel);
            resultView = FindViewById<TextView>(Resource.Id.resultView);
            readButton = FindViewById<Button>(Resource.Id.readButton);
            writeButton = FindViewById<Button>(Resource.Id.writeButton);

            readButton.Click += StartReading;
            writeButton.Click += StartWriting;
            PerformanceTestService.Instance.PiCalculationCompleted += PiCalculationCompleted;

            fileAccessService = new FileAccessTestService(this);

            if (contentToWrite == null)
            {
                progressDialog = ProgressDialog.Show(this, "Przetwarzanie...", "");
                new Thread(new Runnable(() => { PerformanceTestService.Instance.CalculatePi(digits); })).Start();
            }
        }

        private void StartReading(object sender, EventArgs e)
        {
            resultView.Text = "";
            stopwatch = new Stopwatch();
            stopwatch.Start();

            var fileContents = fileAccessService.ReadFromFile(fileName);

            stopwatch.Stop();

            resultView.Text = fileContents;
            timeLabel.Text = stopwatch.GetDurationInMilliseconds();
        }

        private void StartWriting(object sender, EventArgs e)
        {
            resultView.Text = "";
            stopwatch = new Stopwatch();
            stopwatch.Start();

            fileAccessService.WriteToFile(fileName, contentToWrite);

            stopwatch.Stop();
            timeLabel.Text = stopwatch.GetDurationInMilliseconds();
        }

        private void PiCalculationCompleted(string result)
        {
            contentToWrite = result;
            RunOnUiThread(new Runnable(() => { progressDialog.Dismiss(); }));
        }
    }
}

File: Native.Android/app/src/main/java/com/pgrzmil/nativedroid/FileAccessTestActivity.java
package com.pgrzmil.nativedroid;

import android.app.ProgressDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.pgrzmil.services.*;

public class FileAccessTestActivity extends AppCompatActivity implements PerformanceTestListener {
    String fileName = "testFile.txt";
    int digits = 10000;

    Stopwatch stopwatch;
    Button readButton;
    Button writeButton;
    TextView timeLabel;
    TextView resultView;
    ProgressDialog progressDialog;
    FileAccessService fileAccessService;
    String contentToWrite;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fileaccess);
        setTitle(R.string.menuFileAccess);

        timeLabel = (TextView) findViewById(R.id.timeLabel);
        resultView = (TextView) findViewById(R.id.resultView);
        readButton = (Button) findViewById(R.id.readButton);
        writeButton = (Button) findViewById(R.id.writeButton);

        stopwatch = new Stopwatch();
        fileAccessService = new FileAccessService(this);
        PerformanceTestService.getInstance().addListener(this);

        progressDialog = ProgressDialog.show(this, "Przetwarzanie...", "");
        new Thread(new Runnable() {
            public void run() {
                PerformanceTestService.getInstance().calculatePi(digits);
            }
        }).start();
    }

    public void startReading(View view) {
        stopwatch = new Stopwatch();
        stopwatch.start();
        resultView.setText("");

        String fileContents = fileAccessService.readFromFile(fileName);
        resultView.setText(fileContents);

        stopwatch.stop();
        timeLabel.setText(stopwatch.getDurationInMilliseconds());
    }

    public void startWriting(View view) {
        stopwatch = new Stopwatch();
        stopwatch.start();
        resultView.setText("");

        fileAccessService.writeToFile(fileName, contentToWrite);

        stopwatch.stop();
        timeLabel.setText(stopwatch.getDurationInMilliseconds());
    }

    @Override
    public void piCalculationCompleted(String result) {
        contentToWrite = result;
        runOnUiThread(new Runnable() {
            public void run() {
                progressDialog.dismiss();
            }
        });
    }
}
